### PR TITLE
[refactor] Misc improvements over AST helper functions

### LINKED ISSUE #4199: Principled C++ side Expr construction helpers
Following #4119, I searched for similar helper functions and operations for constructing frontend `Expr`s.

- Most operators are in `ir/expression_ops.h`. This file is included by `ir/expression.h`, however not many files are actually using these operators.
- Some others are in `ir/expr.h`
- A few are in `ir/frontend.h`, which appears to be legacy

I propose that we

- Remove the `#include` of `ir/expression_ops.h` from `ir/expression.h` so that new codes have less chance of misusing them. If any file still needs to construct large amount of ASTs they can still include it separately
- Helpers in seem to have two naming conventions (`helper_name` and `HelperName`) - decide on one (I prefer `HelperName`)
- Remove `ir/frontend.h`, move any helpers still in use to `ir/expr.h`

### PR DESCRIPTION
**Pull request by re-xyr**
Related issue = fixes #4199 

- Renamed several PascalCase helpers to snake_case 
- Removed the include of `expression_ops.h` from `expression.h` - any file that needs it should import it individually
- Added `taichi::lang::value<T>(const T &x)` as a version of `taichi::lang::Expr::Expr` that allows specifying types explicitly
- Other misc improvements

### CODE REVIEW

File: taichi/python/export_lang.cpp [hunk 1]
```diff
@@ -266,8 +267,16 @@ void export_lang(py::module &m) {
       .def("begin_frontend_if_true", &ASTBuilder::begin_frontend_if_true)
       .def("pop_scope", &ASTBuilder::pop_scope)
       .def("begin_frontend_if_false", &ASTBuilder::begin_frontend_if_false)
-      .def("insert_deactivate", Deactivate)
-      .def("insert_activate", Activate)
+      .def("insert_deactivate",
+           [](ASTBuilder *ast_builder, SNode *snode,
+              const ExprGroup &expr_group) {
```
Comment (strongoier) [inline taichi/python/export_lang.cpp:272]: We'd better not directly write functions in `export_lang`. Put them inside `ASTBuilder`.

File: taichi/ir/frontend.h [hunk 2]
```diff
@@ -44,59 +39,54 @@ std::vector<Axis> Axes(AX... axes) {
...
-  return dest;
-}
-
 // expr_group are indices
-inline void Activate(ASTBuilder *ast_builder,
-                     SNode *snode,
-                     const ExprGroup &expr_group) {
-  ast_builder->insert(Stmt::make<FrontendSNodeOpStmt>(SNodeOpType::activate,
-                                                      snode, expr_group));
+inline pStmt snode_activate(SNode *snode, const ExprGroup &expr_group) {
```
Comment (strongoier) [inline taichi/ir/frontend.h:43]: I suggest keeping the `insert` behavior and move it inside `ASTBuilder`.

File: taichi/ir/expr.h [hunk 3]
```diff
@@ -134,4 +128,16 @@ Expr bit_cast(const Expr &input) {
   return taichi::lang::bit_cast(input, get_data_type<T>());
 }
 
+// like Expr::Expr, but allows to explicitly specify the type
+template <typename T>
+Expr value(const T &val) {
+  return Expr(val);
+}
+
+Expr expr_rand(DataType dt);
```
Comment (strongoier) [inline taichi/ir/expr.h:137]: Why leave `expr_rand` and `value` here, but put other expressions like `snode_append` in `frontend.h`?

File: taichi/ir/expression_ops.h [hunk 4]
```diff
@@ -2,11 +2,13 @@
...
 #undef DEFINE_EXPRESSION_OP_UNARY
-#undef DEFINE_EXPRESSION_FUNC
+#undef DEFINE_EXPRESSION_OP_BINARY
+#undef DEFINE_EXPRESSION_FUNC_UNARY
+#undef DEFINE_EXPRESSION_FUNC_BINARY
+#undef DEFINE_EXPRESSION_FUNC_TERNARY
 
-#define DEFINE_EXPRESSION_OP_UNARY(opname)                           \
+#define DEFINE_EXPRESSION_FUNC_UNARY(opname)                         \
   Expr opname(const Expr &expr) {                                    \
```
Comment (strongoier) [inline taichi/ir/expression_ops.h:12]: What about ordering them more naturally?

**Review by strongoier — CHANGES_REQUESTED** [review]

File: taichi/ir/expr.h (hunk 3 above)
Comment (re-xyr) [reply taichi/ir/expr.h:137]: done - now `frontend.h` is purged.

**Review by strongoier — APPROVED** [review]
Nice work! Thanks!

**Pull request merged**